pkg/controller/cloudmetrics/cloudmetrics_controller.go

```diff
@@ -77,47 +83,72 @@ var _ reconcile.Reconciler = &ReconcileCloudMetrics{}
 
 // ReconcileCloudMetrics reconciles a CloudMetrics object
 type ReconcileCloudMetrics struct {
-	client client.Client
-	scheme *runtime.Scheme
-	logger *logrus.Entry
+	client             client.Client
+	scheme             *runtime.Scheme
+	logger             *logrus.Entry
+	metricProviderList []providers.MetricsProvider
 }
 
 // Reconcile reads all redis and postgres crs periodically and reconcile metrics for these
 // resources.
 // The Controller will requeue the Request every 5 minutes constantly when RequeueAfter is set
 func (r *ReconcileCloudMetrics) Reconcile(request reconcile.Request) (reconcile.Result, error) {
-	r.logger.Info("Reconciling CloudMetrics")
+	r.logger.Info("reconciling CloudMetrics")
+	ctx := context.TODO()
 
 	// Fetch all redis crs
 	redisInstances := &integreatlyv1alpha1.RedisList{}
 	err := r.client.List(context.TODO(), redisInstances)
 	if err != nil {
 		return reconcile.Result{}, err
 	}
-	if len(redisInstances.Items) > 0 {
-		for _, redis := range redisInstances.Items {
-			r.logger.Infof("Found redis cr: %s", redis.Name)
-		}
-	} else {
-		r.logger.Info("Found no redis instances")
+	for _, redis := range redisInstances.Items {
+		r.logger.Infof("beginning to reconcile cloud metrics for redis cr: %s", redis.Name)
 	}
 
 	// Fetch all postgres crs
 	postgresInstances := &integreatlyv1alpha1.PostgresList{}
 	err = r.client.List(context.TODO(), postgresInstances)
 	if err != nil {
-		return reconcile.Result{}, err
+		r.logger.Error(err)
 	}
-	if len(postgresInstances.Items) > 0 {
-		for _, postgres := range postgresInstances.Items {
-			r.logger.Infof("Found postgres cr: %s", postgres.Name)
+	for _, postgres := range postgresInstances.Items {
+		r.logger.Infof("beginning to reconcile cloud metrics for postgres cr: %s", postgres.Name)
+		if err := r.reconcilePostgresMetrics(ctx, &postgres); err != nil {
+			r.logger.Error(err)
 		}
-	} else {
-		r.logger.Info("Found no postgres instances")
 	}
 
-	// Requeue every 5 minutes
+	// we want full control over when we scrape metrics
+	// to allow for this we only have a single requeue
+	// this ensures regardless of errors or return times
+	// all metrics are scraped and exposed at the same time
 	return reconcile.Result{
-		RequeueAfter: watchDuration * time.Second,
+		RequeueAfter: watchDuration,
 	}, nil
 }
+
+func (r *ReconcileCloudMetrics) reconcilePostgresMetrics(ctx context.Context, postgres *integreatlyv1alpha1.Postgres) error {
+	for _, p := range r.metricProviderList {
+		// only scrape metrics on supported strategies
+		if !p.SupportsStrategy(postgres.Status.Strategy) {
```

Comment:
In relation to my previous comment, I think this isn't going to work if the `r.metricProviderList` contains metric providers for both AWS redis and postgres instances. The AWS Redis Metrics Provider is going to return true for `SupportsStrategy` also.

If you think of the `MetricProvider` interface, it has a matrix of ways it can be implemented. It can be either a Postgres or a Redis and for both of those it can be either OpenShift, AWS, Azure etc. I think it might make more sense to declare interfaces like `PostgresMetricProvider` and `RedisMetricProvider` and then this controller would have two lists like so:

```
type ReconcileCloudMetrics struct {
	client client.Client		client             client.Client
	scheme *runtime.Scheme		scheme             *runtime.Scheme
	logger *logrus.Entry		logger             *logrus.Entry
	postgresMetricProviderList []providers.PostgresMetricsProvider
    redusMetricProviderList    []providers.RedisMetricsProvider
}
```

Reply:
@darahayes Good catch, though the status block of a custom resource has a provider flag. We can just use that instead of the strategy itself. I don't really have a major opinion either way on the multiple list/interface approach.

Reply:
Thinking about this some more, I think the suggestion above is the approach I'd also take if I were writing it. If there's no objections from yourselves @laurafitzgerald @austincunningham @ciaranRoche I think it makes sense to make that change in this PR?